File: mt-pricing-backend/app/schemas/jobs.py
"""Pydantic schemas — Jobs admin (DatabaseScheduler) API contracts.

Alineado con `app/db/models/job.py`. Idioma de campos = inglés (canónico API).

Endpoints expuestos en `app/api/routes/jobs.py`:
- ``GET    /admin/jobs``
- ``GET    /admin/jobs/{id}``
- ``POST   /admin/jobs``           (alta nueva job_definition)
- ``PATCH  /admin/jobs/{id}``      (toggle enabled, editar cron/args/kwargs/queue)
- ``POST   /admin/jobs/{id}/run-now``
- ``GET    /admin/jobs/{id}/runs`` (paginado JobRun)
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

JobOwnerLit = Literal["infra", "business"]
ScheduleTypeLit = Literal["cron", "interval"]
JobStatusLit = Literal["idle", "running", "success", "failure", "cancelled"]


class JobDefinitionListItem(BaseModel):
    """Listado paginado — versión liviana sin args/kwargs."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: str
    task_name: str
    description: str | None = None
    owner: JobOwnerLit
    schedule_type: ScheduleTypeLit
    cron_expression: str | None = None
    interval_seconds: int | None = None
    timezone: str
    queue: str
    enabled: bool
    last_run_at: datetime | None = None
    next_run_at: datetime | None = None
    last_status: JobStatusLit | None = None


class JobDefinitionResponse(JobDefinitionListItem):
    """Detalle completo — incluye args, kwargs, last_error, edited_*."""

    args: list[Any] = Field(default_factory=list)
    kwargs: dict[str, Any] = Field(default_factory=dict)
    last_error: str | None = None
    last_celery_task_id: str | None = None
    edited_by: UUID | None = None
    edited_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class JobDefinitionCreate(BaseModel):
    """Alta nueva job_definition — TI Integración."""

    model_config = ConfigDict(extra="forbid")

    code: Annotated[str, Field(min_length=2, max_length=128)]
    task_name: Annotated[str, Field(min_length=2, max_length=200)]
    description: str | None = Field(default=None, max_length=500)
    owner: JobOwnerLit = "infra"
    schedule_type: ScheduleTypeLit
    cron_expression: str | None = None
    interval_seconds: Annotated[int, Field(ge=1)] | None = None
    timezone: str = "Asia/Dubai"
    queue: str = "default"
    args: list[Any] = Field(default_factory=list)
    kwargs: dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True


class JobDefinitionUpdate(BaseModel):
    """PATCH — sólo los campos que el admin puede editar."""

    model_config = ConfigDict(extra="forbid")

    description: str | None = Field(default=None, max_length=500)
    cron_expression: str | None = None
    interval_seconds: Annotated[int, Field(ge=1)] | None = None
    queue: str | None = None
    args: list[Any] | None = None
    kwargs: dict[str, Any] | None = None
    enabled: bool | None = None


class JobRunResponse(BaseModel):
    """Una ejecución (JobRun) — historial."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    job_id: UUID
    job_code: str
    status: JobStatusLit
    started_at: datetime | None = None
    finished_at: datetime | None = None
    retries: int
    celery_task_id: str | None = None
    result: dict[str, Any] | None = None
    error: str | None = None
    created_at: datetime
    duration_ms: int | None = None


class JobRunsPage(BaseModel):
    """Página de runs."""

    items: list[JobRunResponse]
    count: int
    next_cursor: str | None = None


class JobRunNowResponse(BaseModel):
    """Respuesta de POST /admin/jobs/{id}/run-now."""

    job_id: UUID
    run_id: UUID
    celery_task_id: str | None = None
    enqueued_at: datetime


__all__ = [
    "JobDefinitionCreate",
    "JobDefinitionListItem",
    "JobDefinitionResponse",
    "JobDefinitionUpdate",
    "JobRunNowResponse",
    "JobRunResponse",
    "JobRunsPage",
]

File: mt-pricing-frontend/lib/api/endpoints/jobs.ts
"use client";

import env from "@/lib/env";
import { createSupabaseBrowserClient } from "@/lib/supabase/client";

/**
 * Cliente tipado para `/api/v1/admin/jobs/*` (DatabaseScheduler editable).
 *
 * Backend contracts en `mt-pricing-backend/app/api/routes/jobs.py`.
 */

// ---- Types ----------------------------------------------------------------

export type JobOwner = "infra" | "business";
export type ScheduleType = "cron" | "interval";
export type JobStatus =
  | "idle"
  | "running"
  | "success"
  | "failure"
  | "cancelled";

export interface JobDefinitionListItem {
  id: string;
  code: string;
  task_name: string;
  description: string | null;
  owner: JobOwner;
  schedule_type: ScheduleType;
  cron_expression: string | null;
  interval_seconds: number | null;
  timezone: string;
  queue: string;
  enabled: boolean;
  last_run_at: string | null;
  next_run_at: string | null;
  last_status: JobStatus | null;
}

export interface JobDefinitionDetail extends JobDefinitionListItem {
  args: unknown[];
  kwargs: Record<string, unknown>;
  last_error: string | null;
  last_celery_task_id: string | null;
  edited_by: string | null;
  edited_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface JobDefinitionCreatePayload {
  code: string;
  task_name: string;
  description?: string | null | undefined;
  owner?: JobOwner | undefined;
  schedule_type: ScheduleType;
  cron_expression?: string | null | undefined;
  interval_seconds?: number | null | undefined;
  timezone?: string | undefined;
  queue?: string | undefined;
  args?: unknown[] | undefined;
  kwargs?: Record<string, unknown> | undefined;
  enabled?: boolean | undefined;
}

export interface JobDefinitionUpdatePayload {
  description?: string | null | undefined;
  cron_expression?: string | null | undefined;
  interval_seconds?: number | null | undefined;
  queue?: string | undefined;
  args?: unknown[] | undefined;
  kwargs?: Record<string, unknown> | undefined;
  enabled?: boolean | undefined;
}

export interface JobRun {
  id: string;
  job_id: string;
  job_code: string;
  status: JobStatus;
  started_at: string | null;
  finished_at: string | null;
  retries: number;
  celery_task_id: string | null;
  result: Record<string, unknown> | null;
  error: string | null;
  created_at: string;
  duration_ms: number | null;
}

export interface JobRunsPage {
  items: JobRun[];
  count: number;
  next_cursor: string | null;
}

export interface JobRunNowResponse {
  job_id: string;
  run_id: string;
  celery_task_id: string | null;
  enqueued_at: string;
}

// ---- Internals ------------------------------------------------------------

export class JobsApiError extends Error {
  public readonly status: number;
  public readonly detail: unknown;

  constructor(status: number, detail: unknown, fallback: string) {
    super(typeof detail === "string" ? detail : fallback);
    this.name = "JobsApiError";
    this.status = status;
    this.detail = detail;
  }
}

async function authedFetch<T>(path: string, init: RequestInit = {}): Promise<T> {
  const supabase = createSupabaseBrowserClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  const headers = new Headers(init.headers);
  if (!headers.has("Content-Type") && init.body && !(init.body instanceof FormData)) {
    headers.set("Content-Type", "application/json");
  }
  if (session?.access_token) {
    headers.set("Authorization", `Bearer ${session.access_token}`);
  }
  const res = await fetch(`${env.NEXT_PUBLIC_BACKEND_URL}${path}`, {
    ...init,
    headers,
    cache: "no-store",
  });
  if (!res.ok) {
    let detail: unknown;
    try {
      detail = await res.json();
    } catch {
      /* noop */
    }
    throw new JobsApiError(res.status, detail, res.statusText);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

function buildQuery(
  params: Record<string, string | number | boolean | undefined | null>,
): string {
  const search = new URLSearchParams();
  Object.entries(params).forEach(([k, v]) => {
    if (v !== undefined && v !== null) search.set(k, String(v));
  });
  const s = search.toString();
  return s ? `?${s}` : "";
}

// ---- API ------------------------------------------------------------------

export const jobsAdminApi = {
  list: (params: {
    enabled?: boolean | undefined;
    owner?: JobOwner | undefined;
    limit?: number | undefined;
    offset?: number | undefined;
  } = {}) =>
    authedFetch<JobDefinitionListItem[]>(
      `/api/v1/admin/jobs${buildQuery(params)}`,
    ),
  get: (id: string) =>
    authedFetch<JobDefinitionDetail>(`/api/v1/admin/jobs/${id}`),
  create: (payload: JobDefinitionCreatePayload) =>
    authedFetch<JobDefinitionDetail>(`/api/v1/admin/jobs`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  update: (id: string, payload: JobDefinitionUpdatePayload) =>
    authedFetch<JobDefinitionDetail>(`/api/v1/admin/jobs/${id}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),
  runNow: (id: string) =>
    authedFetch<JobRunNowResponse>(`/api/v1/admin/jobs/${id}/run-now`, {
      method: "POST",
    }),
  listRuns: (
    id: string,
    params: { limit?: number | undefined; offset?: number | undefined } = {},
  ) =>
    authedFetch<JobRunsPage>(
      `/api/v1/admin/jobs/${id}/runs${buildQuery(params)}`,
    ),
};
